velox/functions/sparksql/DateTimeFunctions.h

```diff
@@ -209,4 +209,21 @@ struct LastDayFunction {
   }
 };
 
+template <typename T>
+struct DateSubFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Date>& date,
+      const int64_t value) {
```

Comment:
Why is this int64_t? Shouldn't it be int32_t? 

Please, clarify.

Reply:
Hi Basmanova，
Used int64_t to check the overflow of int32_t in case of user input unexpected very large value. 
But the actually valid input value is int32_t. 
See below function implementation. 
Kindly let me know if any comments. Thanks
    //check overflow
    if (value != (int32_t)value) {
      VELOX_UNSUPPORTED("integer overflow");
    }
	//actually valid int32_t input value
    int32_t subValue = 0 - (int32_t)value;
    result = addToDate(date, DateTimeUnit::kDay, subValue);

Reply:
I think the signature should use int32_t, but the original question I had was about what happens if users pass std::numeric_limits<int32_t>::min/max? Do these work?

Reply:
Hi Basmanova，
Thanks a lot for your patient review.
I didn't understand your question initially. Yes, we should use int32_t. 
I have tried the min/max of <int32_t> test case.  min date is "-5877641-06-25", max date is "5881580-07-10".
min/max value -/+ 2147483647 works. Kindly refer to below logs.
I didn't add the max value test case yet, it's too huge. ^_^
date_sub spark function calls presto addToDate function, presto date_add also use it. if need to add protection, it's better to add in addToDate function.
Kindly let me know if any further comments. 
Have a good weekend.

//Below is boundary of the success test cases.
EXPECT_EQ(parseDate("5881580-07-10"), dateSub("1969-12-31", -2147483647));
EXPECT_EQ(parseDate("-5877641-06-25"), dateSub("1970-01-02", 2147483647));

//When exceed the boundary, below test case failed. 
 EXPECT_EQ(parseDate("-5877641-06-25"), dateSub("1969-01-01", 2147483647));
 EXPECT_EQ(parseDate("5881580-07-10"), dateSub("1970-12-31", -2147483647));
 ../../velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:231: Failure
Expected equality of these values:
  parseDate("-5877641-06-25")
    Which is: -2147483646
  dateSub("1969-01-01", 2147483647)
    Which is: (2147483284)
../../velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:232: Failure
Expected equality of these values:
  parseDate("5881580-07-10")
    Which is: 2147483646
  dateSub("1970-12-31", -2147483647)
    Which is: (-2147483285)

Reply:
> I have tried the min/max of <int32_t> test case. min date is "-5877641-06-25", max date is "5881580-07-10".
min/max value -/+ 2147483647 works. Kindly refer to below logs.

Thank you for checking. Do these results match Spark?

Reply:
> I didn't add the max value test case yet, it's too huge. ^_^

I'm not sure I understand. What is "too huge"? Any chance you could elaborate?

Reply:
Sorry for confusion , I mean the numbers is so huge. 
I am not sure if I should add these test cases? likes below.
What's your suggestion?
  EXPECT_EQ(parseDate("2019-02-28"), dateSub("2019-03-01", 1));
  EXPECT_EQ(parseDate("2019-01-30"), dateSub("2020-02-29", 395));
  EXPECT_EQ(parseDate("2020-02-29"), dateSub("2019-02-28", -366));
  EXPECT_EQ(parseDate("1970-01-02"), dateSub("-5877641-06-25", -2147483647));
  EXPECT_EQ(parseDate("1969-12-31"), dateSub("5881580-07-10", 2147483647));
  EXPECT_EQ(parseDate("5881580-07-10"), dateSub("1969-12-31", -2147483647));
  EXPECT_EQ(parseDate("-5877641-06-25"), dateSub("1970-01-02", 2147483647));

Reply:
Hi @mbasmanova,
Have confirmed the results matched with Spark. Test cased added also. Kindly help to review again. Thanks.